onnxruntime/core/providers/mkldnn/activation/activations.cc

```diff
@@ -0,0 +1,223 @@
+	GetPrimitiveDesc() const {
+		return context_.relu_fwd_pd;
+  }
+
+ private:
+  struct ReluContext {
+    mkldnn::memory::format src_fmt;
+    mkldnn::memory::format dst_fmt;
+
+    std::unique_ptr<mkldnn::memory> src_mem;
+    std::unique_ptr<mkldnn::memory> dst_mem;
+
+    std::unique_ptr<mkldnn::eltwise_forward::desc> fwd_desc;
+    std::unique_ptr<mkldnn::eltwise_forward::primitive_desc> relu_fwd_pd;
+    std::unique_ptr<mkldnn::primitive> relu_fwd;
+
+    std::unique_ptr<mkldnn::memory::desc> src_md;
+    std::unique_ptr<mkldnn::memory::desc> dst_md;
+
+    std::unique_ptr<mkldnn::stream> stream;
+    std::vector<mkldnn::primitive> net;
+
+    ReluContext()
+    : src_fmt(mkldnn::memory::format::any),
+      dst_fmt(mkldnn::memory::format::any),
+      src_mem(nullptr),
+      dst_mem(nullptr),
+      fwd_desc(nullptr),
+      relu_fwd_pd(nullptr),
+      relu_fwd(nullptr),
+      src_md(nullptr),
+      dst_md(nullptr),
+      stream(nullptr) {}
+  };
+
+  void Initialize(const ReluParams& params) {
+    
+    mkldnn::memory::format fmt = mkldnn::memory::format::any;
+    switch (params.num_dimensions) {
+    case 1: { fmt = mkldnn::memory::format::x; break; }
+    case 2: { fmt = mkldnn::memory::format::nc; break; }
+    case 3: { fmt = mkldnn::memory::format::ntc; break; }
+    case 4: { fmt = mkldnn::memory::format::nchw; break; }
+    case 5: { fmt = mkldnn::memory::format::ncdhw; break; }
+    default: {  fmt = mkldnn::memory::format::any; break; }
+    }
+
+    context_.src_md.reset(new mkldnn::memory::desc(
+      { params.src_dims }, MklDnnType<T>(), fmt));
+    context_.dst_md.reset(new mkldnn::memory::desc(
+      { params.dst_dims }, MklDnnType<T>(), fmt));
+
+    context_.fwd_desc.reset(new mkldnn::eltwise_forward::desc(
+      mkldnn::prop_kind::forward_inference, mkldnn::algorithm::eltwise_relu, 
+      *context_.src_md, 0, 0));
+
+    context_.relu_fwd_pd.reset(
+      new mkldnn::eltwise_forward::primitive_desc(*context_.fwd_desc,
+      cpu_engine_));
+
+    context_.src_fmt = static_cast<mkldnn::memory::format>(
+      context_.relu_fwd_pd.get()->dst_primitive_desc().desc().data.format);
+
+    context_.dst_fmt = static_cast<mkldnn::memory::format>(
+      context_.relu_fwd_pd.get()->dst_primitive_desc().desc().data.format);
+
+    context_.src_mem.reset(
+     new mkldnn::memory(context_.relu_fwd_pd.get()->dst_primitive_desc(), 
+			  nullptr));
+    context_.dst_mem.reset(
+      new mkldnn::memory(context_.relu_fwd_pd.get()->dst_primitive_desc(), 
+			nullptr));
+    context_.relu_fwd.reset(
+      new mkldnn::eltwise_forward(*context_.relu_fwd_pd, *context_.src_mem, 
+			*context_.dst_mem));
+    context_.net.push_back(*context_.relu_fwd);
+  }
+
+  ReluContext context_;
+  mkldnn::engine& cpu_engine_;
+};
+
+// Pool which allows for reuse of MKLDNN Relu primitives which are expensive 
+// to instantiate. To address thread safety, the primitives are stored in a map 
+// on thread local storage.
```

Comment:
Who is in the thread local storage?

Reply:
@snnn kernel objects like Relu lives in thread local storage. This is very similar to what @weixingzhang's implementation of Conv and Pool kernels.